contracts/account/extensions/ERC7579Modules/SocialRecoveryExecutor.sol

```diff
@@ -0,0 +1,475 @@
+
+    /// @notice Computes the EIP-712 digest (typed data hash) for a recovery request
+    /// @dev Uses the current nonce to prevent replay attacks across recovery attempts
+    /// @param account The account address for which the recovery digest is being generated
+    /// @return bytes32 The EIP-712 compliant digest
+    function _getRecoveryDigest(address account) internal view returns (bytes32) {
+        bytes32 structHash = keccak256(abi.encode(RECOVERY_TYPEHASH, account, getNonce(account)));
+        return _hashTypedDataV4(structHash);
+    }
+
+    /// @notice Verifies multiple guardian signatures
+    /// @dev Checks each signature against the current recovery digest
+    /// @param account The account the signatures are for
+    /// @param guardianSignatures Array of guardian signatures to verify
+    /// @return True if all signatures are valid, false otherwise
+    function guardianSignaturesAreValid(
+        address account,
+        GuardianSignature[] calldata guardianSignatures
+    ) public view returns (bool) {
+        if (guardianSignatures.length > 32) {
+            revert MaxGuardians();
+        }
+
+        for (uint256 i = 0; i < guardianSignatures.length; i++) {
+            // check signature validity
+            if (!guardianSignatureIsValid(account, guardianSignatures[i])) {
+                return false;
+            }
+            // check for signature duplication @TBD optimize O(n^2)
+            address currentSigner = guardianSignatures[i].signer;
+            for (uint256 j = 0; j < i; j++) {
+                if (guardianSignatures[j].signer == currentSigner) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    /// @notice Verifies a single guardian signature against the current recovery digest
+    /// @dev Checks if the signature is valid and if the signer is a guardian
+    /// @param account The account the signature is for
+    /// @param guardianSignature The signature to verify
+    /// @return true if the signature is valid, false otherwise
+    function guardianSignatureIsValid(
+        address account,
+        GuardianSignature calldata guardianSignature
+    ) public view returns (bool) {
+        return
+            isGuardian(account, guardianSignature.signer) &&
+            SignatureChecker.isValidSignatureNow(
+                guardianSignature.signer,
+                _getRecoveryDigest(account),
+                guardianSignature.signature
+            );
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                MODULE MANAGEMENT
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Adds a new guardian to the account's recovery configuration
+    /// @dev Can only be called by the account itself
+    /// @param guardian Address of the new guardian
+    function addGuardian(address guardian) external {
```

Comment:
this logic is duplicated in `onInstall`.

You should probably have an internal `function _addGuardian(address account, address guardian) internal`

That you could call from here and from `onInstall`

Reply:
In general, we try to always provide internal virtual version of all these function (with account as a parameter), and then have external functions call the internal ones with msg.sender.